source/common/network/utility.cc

```diff
@@ -249,8 +249,9 @@ bool Utility::isLocalConnection(const Network::ConnectionSocket& socket) {
   // while the client is not actually local. Example could be an iptables intercepted
   // connection. However, this is a rare exception and such assumption results in big
   // performance optimization.
-  if (remote_address->type() == Envoy::Network::Address::Type::Pipe ||
-      remote_address == socket.localAddress() || isLoopbackAddress(*remote_address)) {
+  if (remote_address->type() != Envoy::Network::Address::Type::Ip ||
+      isSameIPAddress(*remote_address, *socket.localAddress()) ||
```

Comment:
Sorry, but could you explain why doesn't this work with `==`?

Also, can't we fix the `operator==` instead of adding an utility function, since presumably this is broken in other places as well?

Reply:
`remote_address` (as well as `socket.localAddress` are shared pointers (`InstanceConstSharedPtr`). Equality operator of pointers compares the pointers themselves, rather than the objects they point to.

Secondly, even if the addresses were compared, this comparison must ignore the port numbers that are embedded in every Envoy `Ip` address instance.

Reply:
`operator==` is correct for comparing the whole IP address + port combination. It also works in many places where it is known that the port numbers of both addresses are the same (e.g., zero).

Reply:
You may want to review the two commits separately, as they address the above issues one at a time.

Reply:
Ah, I completely missed the port part, thanks!